### PR TITLE
Multilanguage (but not only): Adding field menuitem to login menu item

### PR DESCRIPTION
**Pull request by infograf768**
For a reason that I ignore (history I guess), the Login menu item contains a Redirect field in text format.
For multilanguage (as well as monolanguage), it is better to have a "menuitem" type field, which will prevent any error as the redirect should be internal. In that case it is an Itemid which is saved in the database (as already exists for the Logout menu item).
The issue here was to remain B/C while letting people use the alternative "menuitem" field.

To accomplish this, I added a new field of type "menuitem" (for the login and logout parts of the Login menu item), created 2 new rules, and added showon to let display only one field at a time and throw an alert preventing saving a value both in the ancient "Login Redirect"  and the new "Menu Item Login Redirect" and as well for "Logout Redirect" and the new "Menu Item Logout Redirect".

When the menuitem field is used, the patch also checks if Multilang is on and, if it is, adds the necessary lang tag to the redirect.

_Once this is merge, we will be able to merge a pending PR https://github.com/joomla/joomla-cms/pull/9724 allowing multilanguage users to choose any redirection in the login module too, including associations when automatic change is on._

New interface (it is B/C as default is "Login Redirect" text field)

screen shot 2016-05-05 at 09 48 12

If choosing "Menu Item"

screen shot 2016-05-05 at 09 49 26

To test, enter internal values in the "Login redirect" and "Logout redirect" fields, save.
Then patch and test in monolanguage and multilanguage.

For example, if both fields are used for login (or/and logout), validation will prevent saving the menu item and we will get a Warning.

screen shot 2016-05-05 at 09 53 43

or

screen shot 2016-05-05 at 09 56 03

### CODE REVIEW

**Comment by infograf768** [comment]
@andrepereiradasilva @brianteeman @richard67 @Bakual @mannybiker 

Thanks for testing.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 6fa9124061a4749b46ef5097688863a5c99eb593<br /><br />Great stuff<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10249">issues.joomla.org/joomla-cms/10249</a>.</sub>

**Comment by brianteeman** [comment]
Is it possible to make the default be the new menu item option and the custom url only displayed on load if it is set
that should still be b/c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10249">issues.joomla.org/joomla-cms/10249</a>.</sub>

**Comment by infograf768** [comment]
@brianteeman
If we set the "Choose Login Redirect Type"  field default to Menu Item (i.e. value="1", it will indeed not kill what was entered before in the present "Login Redirect" field, thus being B/C indeed. It may just confuse the user who already used that field.
It is an easy change. Let's see what other testers think.

**Comment by brianteeman** [comment]
But can it be set so that IF custom url not blank then it is displayed.
That should remove any confusion if possible

On 5 May 2016 at 09:21, infograf768 <notifications@github.com> wrote:

> @brianteeman <https://github.com/brianteeman>
> If we set the "Choose Login Redirect Type" field default to Menu Item
> (i.e. value="1", it will indeed not kill what was entered before in the
> present "Login Redirect" field, thus being B/C indeed. It may just confuse
> the user who already used that field.
> It is an easy change. Let's see what other testers think.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/10249#issuecomment-217101271>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by infograf768** [comment]
let me see

**Comment by infograf768** [comment]
nope, can'f find how to.
needs more tests and feedback

**Comment by mannybiker** [comment]
Item tested and working. Very useful addition since it was very time consuming and reason for mistakes to find all the correct article id, catid, itemid to compose manually the URL.

I agree with brianteeman on the default behaviour of having the menu item as default option if the custom URL filed is empty, but I understand that if this is technically difficult to achieve it is probably best to keep the Custom URL as the first option so that it is clear that either there is no configured URL or the old data is still there.

**Comment by infograf768** [comment]
Found the solution. Easier in the morning 😄 
The custom URL redirect fields will now display when the fields have a value. I obtained this result by adding a special case in the getItem() method in ROOT/administrator/components/com_menus/models/item.php
Menu Item is now default (test creating a new login menu item).
(Also took off a useless variable in the new rules files)

@mannybiker 
@andrepereiradasilva 
@brianteeman 

Thanks for testing

**Comment by brianteeman** [comment]
Awesome - knew you wouldnt give up until you worked out how to do it

On 6 May 2016 at 07:53, infograf768 <notifications@github.com> wrote:

> Found the solution. Easier in the morning 😄
> The custom URL redirect fields will now display when the fields have a
> value. I obtained this result by adding a special case in the getItem()
> method in ROOT/administrator/components/com_menus/models/item.php
> Menu Item is now default (test creating a new login menu item).
> (Also took off a useless variable in the new rules files)
>
> @mannybiker <https://github.com/mannybiker>
> @andrepereiradasilva <https://github.com/andrepereiradasilva>
> @brianteeman <https://github.com/brianteeman>
>
> Thanks for testing
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/10249#issuecomment-217362382>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by infograf768** [comment]
Can you folks set this RTC in issues (after testing 😃 )

**Comment by brianteeman** [comment]
Dont worry - I'm on a train to the sprint right now but will make sure it
gets merged over the weekend

On 6 May 2016 at 08:07, infograf768 <notifications@github.com> wrote:

> Can you folks set this RTC in issues (after testing 😃 )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/10249#issuecomment-217365934>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

File: administrator/language/en-GB/en-GB.com_users.ini [hunk 1]
```diff
@@ -92,6 +92,17 @@ COM_USERS_ERROR_VIEW_LEVEL_IN_USE="You can't delete the view access level '%d:%s
 COM_USERS_ERROR_SECRET_CODE_WITHOUT_TFA="You have entered a Secret Code but two factor authentication is not enabled in your user account. If you want to use a secret code to secure your login please edit your user profile and enable two factor authentication."
 COM_USERS_FIELD_CATEGORY_ID_LABEL="Category"
 COM_USERS_FIELD_ID_LABEL="ID"
+COM_USERS_FIELD_LOGIN_MENUITEM="Menu Item"
+COM_USERS_FIELD_LOGIN_REDIRECT_CHOICE_DESC="'Custom URL' lets enter any internal url in the Redirect field while 'Menu Item' let's directly select an existing menu item.<br />For a multilingual site, it is advised to choose 'Menu Item'."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_users.ini]: Can you change this string to the following as it is better English

+COM_USERS_FIELD_LOGIN_REDIRECT_CHOICE_DESC="'Custom URL' lets you manually enter any internal url in the Redirect field. 'Menu Item' lets you directly select an existing menu item.<br />For a multilingual site, it is advised to use 'Menu Item'."

**Comment by brianteeman** [comment]
Made a small suggestion for one of the language strings<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10249">issues.joomla.org/joomla-cms/10249</a>.</sub>

**Comment by infograf768** [comment]
Done. Thanks Brian.

**Comment by brianteeman** [comment]
Thanks

**Comment by andrepereiradasilva** [comment]
Didn't do a full test, but discovered this issue.
Used sample data. Created a login menu item with this values.

image

Custom URL in login didn't work, the custom URL in logout worked.

**Comment by andrepereiradasilva** [comment]
Ok i see we have to add internal URI (`index.php?....`)... hum, this should be in the tooltip.

**Comment by MATsxm** [comment]
I have tested this item :white_check_mark: successfully on 5265043805ab192311130315649f8351b117a889<br /><br />Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10249">issues.joomla.org/joomla-cms/10249</a>.</sub>

**Comment by infograf768** [comment]
@andrepereiradasilva 
the tip for the redirect url field is in main en-GB.ini. JFIELD_LOGIN_ etc. I did not touch it. It is unrelated to this PR

**Comment by andrepereiradasilva** [comment]
> the tip for the redirect url field is in main en-GB.ini. JFIELD_LOGIN_ etc. I did not touch it. It is unrelated to this PR

i know, just pointing out that.

Tests in en-GB ok. Still have to test in a multilanguage env.

**Comment by infograf768** [comment]
@andrepereiradasilva 
I agree the tips for both redirect url are no good as they only say to enter an internal url without specifying the kind of format expected. (Which has created issues when we changed the code in 3.4.8 as the redirect should now be a  JURI one.)
My proposal would be to let alone the constant and value in the main ini file and maybe change here all constants of new ones of type COM_USERS_ with better tip, and, as you said in our private chat, add in both fields a placeholder.
OR we could just add the placeholder.
What do you think?

**Comment by andrepereiradasilva** [comment]
i would add the new tooltip indicating that AND add placeholders (in all login/logout forms). But in another PR.

**Comment by infograf768** [comment]
OK, let's do that in another PR and get this one in first.

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 5265043805ab192311130315649f8351b117a889<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10249">issues.joomla.org/joomla-cms/10249</a>.</sub>

**Comment by brianteeman** [comment]
Rtc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10249">issues.joomla.org/joomla-cms/10249</a>.</sub>

**Comment by roland-d** [comment]
Merged, thanks everybody for participating.

**Comment by infograf768** [comment]
thanks @roland-d

**Comment by infograf768** [comment]
Strings modified here: https://github.com/joomla/joomla-cms/pull/10336
please test

**Pull request merged**